sdk/core/Azure.Core/src/Pipeline/Internal/RequestScopedHttpMessageProperties.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Azure.Core.Pipeline
+{
+    internal static class RequestScopedHttpMessageProperties
+    {
+        private static readonly AsyncLocal<HttpMessagePropertiesScope?> CurrentHttpMessagePropertiesScope = new AsyncLocal<HttpMessagePropertiesScope?>();
+
+        internal static IDisposable StartScope(string name, object value)
+        {
+            CurrentHttpMessagePropertiesScope.Value = new HttpMessagePropertiesScope(name, value, CurrentHttpMessagePropertiesScope.Value);
+
+            return CurrentHttpMessagePropertiesScope.Value;
+        }
+
+        internal static void AddHttpMessageProperties(HttpMessage message)
+        {
+            if (CurrentHttpMessagePropertiesScope.Value != null)
+            {
+                foreach (KeyValuePair<string, object> kvp in CurrentHttpMessagePropertiesScope.Value.Properties)
```

Comment:
We have an option of copying values from parents or just traversing the hierarchy here.

Reply:
I was thinking about traversal but I think the logic would be more complicated unless we're ok with recursion (i.e. we have to apply parent value before child). The traversal would have smaller memory footprint I think but on the other hand I don't expect many scenarios with nested scopes. What would you prefer?

Reply:
Let's keep the dictionary-based logic you have.